### PR TITLE
PLANNER-2541 Simplify ProblemFactChange interface

### PR DESCRIPTION
**Pull request by rsynek**
### JIRA
https://issues.redhat.com/browse/PLANNER-2541

### Referenced pull requests
https://github.com/kiegroup/optaplanner/pull/1703
https://github.com/kiegroup/optaplanner-quickstarts/pull/259
https://github.com/kiegroup/optaweb-vehicle-routing/pull/635

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add the label `run_fdb`
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] tests</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner-quickstarts] native</b>
</details>

### CODE REVIEW

File: use-cases/call-center/src/main/java/org/acme/callcenter/service/SolverService.java [hunk 1]
```diff
@@ -62,13 +62,13 @@ public SolverService(SolverFactory<CallCenter> solverFactory, @Default ManagedEx
...
-        List<ProblemFactChange<CallCenter>> pinCallProblemFactChanges = calls.stream()
+        List<ProblemChange<CallCenter>> pinCallProblemChanges = calls.stream()
                 .filter(call -> !call.isPinned()
                         && call.getPreviousCallOrAgent() != null
                         && call.getPreviousCallOrAgent() instanceof Agent)
-                .map(PinCallProblemFactChange::new)
+                .map(PinCallProblemChange::new)
                 .collect(Collectors.toList());
-        solver.addProblemFactChanges(pinCallProblemFactChanges);
+        solver.addProblemChanges(pinCallProblemChanges);
```
Comment (triceo) [inline use-cases/call-center/src/main/java/org/acme/callcenter/service/SolverService.java]: Same comment as earlier - you're building a list, just to pass it somewhere to be unrolled again. Why not do a `.map(...).forEach(...)` instead?

File: use-cases/call-center/src/main/java/org/acme/callcenter/solver/change/RemoveCallProblemChange.java [hunk 2]
```diff
@@ -16,39 +16,36 @@
...
 
     @Override
-    public void doChange(ScoreDirector<CallCenter> scoreDirector) {
-        CallCenter callCenter = scoreDirector.getWorkingSolution();
+    public void doChange(CallCenter workingCallCenter, ProblemChangeDirector problemChangeDirector) {
         Call call = new Call(callId, null);
-        Call workingCall = scoreDirector.lookUpWorkingObjectOrReturnNull(call);
-        if (workingCall != null) {
+        Optional<Call> workingCallOptional = problemChangeDirector.lookUpWorkingObjectOptionally(call);
+        workingCallOptional.ifPresent(workingCall -> {
```
Comment (triceo) [inline use-cases/call-center/src/main/java/org/acme/callcenter/solver/change/RemoveCallProblemChange.java]: I'd consider extracting this lambda to its own method.

**Review by triceo — APPROVED** [review]
LGTM, minor comments.

File: use-cases/call-center/src/main/java/org/acme/callcenter/solver/change/RemoveCallProblemChange.java (hunk 2 above)
Comment (rsynek) [reply use-cases/call-center/src/main/java/org/acme/callcenter/solver/change/RemoveCallProblemChange.java]: That would look much better, I agree. But unfortunately, the lambda needs to access the `ProblemChangeDirector` and the working solution.

Comment (triceo) [reply use-cases/call-center/src/main/java/org/acme/callcenter/solver/change/RemoveCallProblemChange.java]: And it still could. :-) 
`workingCall -> callSomething(pcd, workingsolution)`

Comment (rsynek) [reply use-cases/call-center/src/main/java/org/acme/callcenter/solver/change/RemoveCallProblemChange.java]: Ah yes, thanks. I was thinking about putting there a method reference, which would be really nice.

**Pull request merged**